### DISCUSSION TITLE
F# resources display support

### DISCUSSION THREAD

**Discussion by vzarytovskii** [codemerx/CodemerxDecompile#103] [Q&A] [answered]
Hey folks, would you accept a contribution  to be able to display F# resources?

Explanation: F# compiler emits additional assembly resources, which contain some optimisation and type data, it's very helpful sometimes when we investigate codegen, and would be awesome if it can be in instrument like this.

**Comment by tsviatko** [comment]
Hey @vzarytovskii we will be glad to have a  feature that allows us to display F# resources.

I really hope you have not started implementing it already, though. By some weird coincidence just yesterday when you posted this issue  we released a new version of CodemerxDecompile with completely rewritten UI. We switched from VS Code clone to custom Avalonia UI based UI. So if you already have some implementation against the old VS Code UI there will be collisions. 

So that might be a hiccup. Other than that we will be really glad to have some extra functionality that might be useful to the F# developer community. On a side note, there is a  chance that we add F# decompilation at some point in the future since we have been getting some requests for it.

**Reply by vzarytovskii** [reply]
I haven't started with it yet, was just exploring, thanks for the heads up

**Reply by tsviatko** [reply]
Ah OK. Then now is a really good time to start :D 

Actually maybe we can help with that too. We used to have assembly resource reader in JustDecompile. It did not qualify for open sourcing, though. So we have it planned but we have not gotten to it yet. Maybe we can work on this together so that we can reuse common parts.

For starters, could you send over like a sample assembly that has the resources you would like to see? That would allow us to investigate whether the F# compiler resource embedding scheme has some specifics that we are not aware of.

**Reply by vzarytovskii** [reply]
> Ah OK. Then now is a really good time to start :D 
> 
> 
> 
> Actually maybe we can help with that too. We used to have assembly resource reader in JustDecompile. It did not qualify for open sourcing, though. So we have it planned but we have not gotten to it yet. Maybe we can work on this together so that we can reuse common parts.
> 
> 
> 
> For starters, could you send over like a sample assembly that has the resources you would like to see? That would allow us to investigate whether the F# compiler resource embedding scheme has some specifics that we are not aware of.

I can for sure, they also available on nuget (for example https://www.nuget.org/packages/FSharp.Core). There are nothing special about how they're stored (they are just normal assembly resources), however encoding of those are specific to F# compiler, and this is something I can for sure help with.

**Reply by tsviatko** [reply]
This is interesting. Does the encoding change between the versions of the F# compiler? Is there some kind of public spec on this or is it an internal implementation detail??

**Reply by vzarytovskii** [reply]
It doesn't change "much", though it can, but it's possible to implement defensive "unpicking" of the data, where unknown values won't be converted from byte array to specific type. Format is an implementation detail, and is deliberately not standardised anywhere.

**Reply by tsviatko** [reply, marked as answer]
OK I created an issue for displaying assembly resource. Once we have this working we could build the F# specific stuff on top. If you guys start some implementation beforehand, please, let us know.

**Reply by vzarytovskii** [reply]
Yeah, I can just directly port reader/unpacker from compiler from F# to C# when reading is ready.
